### ISSUE TITLE
Create new matcher `toBeLabelled`

### ISSUE THREAD

**Issue by hiwelo** [testing-library/jest-dom#111]
### Describe the feature you'd like:

The idea is to have a custom matcher on the model of `toBeRequired` or `toBeDisabled` helping folks to know if they labelled correctly the markup they are testing.

This helper would be interesting for me because there is a variety of ways to label correctly an element, such as:

* `aria-label`
* `aria-labelledby`
* `alt` on images
* `title` on images and SVGs
*  having a text content (by using `.toHaveTextContent`)

The main limitation is that I am not sure we can easily know when a DOM node is labelled thanks to a nearby _visually-hidden_ or displayed text.

### Suggested implementation:

Just doing all the tests decribed few lines before.

### Describe alternatives you've considered:

Doing each test one after the other -or- using `jest-axe` but in this case lacking flexibility on doing more granular tests.

### Teachability, Documentation, Adoption, Migration Strategy:

* [WCAG 2.0 G94: Providing short text alternative for non-text content that serves the same purpose and presents the same information as the non-text content](https://www.w3.org/TR/WCAG20-TECHS/G94.html)
* https://www.sarasoueidan.com/blog/accessible-icon-buttons/
* https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_aria-label_attribute
* https://www.w3.org/TR/wai-aria/#aria-label

I could help with the implementation if you're up to adding this to this library :+1:

**Comment by gnapse** [comment]
I like this!

About this:

> The main limitation is that I am not sure we can easily know when a DOM node is labelled thanks to a nearby visually-hidden or displayed text which is not part of the same parent (but in this case, it would also be complex to find for screen readers/assistive technologies).

If I'm reading correctly, these cases are not really examples of something that's properly labelled, aren't they. So it's ok if this new matcher would fail on them.

**Comment by alexkrolick** [comment]
> This helper would be interesting for me because there is a variety of ways to label correctly an element, such as:
> 
> [...]

That's a lot of different concepts lumped together... not sure they all fit map well enough onto the concept of "label" especially given the dom-testing queries are separated for alt-text and label/aria-label.

**Comment by gnapse** [comment]
I agree that image alt text should be included in the mix. It's a different concept. Otherwise you might end up with this situation:

```html
<img aria-label="Dog" src="dog.jpg" />
```

```js
expect(imageElement).toBeLabelled() // <- true for the above image element
```

when in reality that image does not have the one thing it should have, the `alt` attribute.

But I do like the idea of having this matcher and have it cover checking that the given element has to have either `aria-label` or `aria-labelledby`, or be labelled by its text content. That should be enough and of great help. It would allow to write tests that are not broken if the implementation decides to change the way in which a label for an element is provided.

**Comment by hiwelo** [comment]
That is an interesting question and I am not sure the whole accessibility community really settled on.

First, the W3C is providing an [Accessible name and description calculation method](https://www.w3.org/TR/html-aapi/#accessible-name-and-description-calculation) for a series of HTML elements. For images, the W3C is giving [this calculation method](https://www.w3.org/TR/html-aapi/#h3_img-element):

1. `aria-labelledby`
2. `aria-label`
3. `alt`
4. `title`

For this reason, I would say that an image can have an accessible name even if there is no `alt` attribute. It is not the most conventional thing to do, but basically assistive technologies have a preference for `aria-*` attribute over the rest of an element's properties. So if an image is having both `alt` and `aria-label`, the `aria-label` would have precedence.

So yes, there is a clear "conflict" between the [HTML 5.1 recommendations](https://www.w3.org/TR/html51/) and the [Accessibility API implementation guide](https://www.w3.org/TR/html-aapi/). 
For me the question is what do we want to check here:

* are we checking if the image is having a label/accessible name? In this case, I would say that `<img aria-label="Description" src="" />` is having an accessible name, even if the markup is not W3C HTML5.1 compliant.
* are we checking if the image is having a valid HTML markup and a valid text alternative? But in this case, is it really the goal of this package?

For me, `jest-dom` is more about checking the current state of something in the DOM than checking the validity of it. But it's just an opinion, not really an argument. 

But yeah, all of this is for me the main reason why all the checks described earlier are still useful, even on images. 🙂 
What are you thoughts around that @alexkrolick @gnapse?

**Comment by hiwelo** [comment]
Also, just to be clearer: I consider `<img src="" aria-label="Label" />` as a bad practice. An image must have an `src` attribute.
I am just wondering what should be the behaviour of this specific check.

Imho, checking the validity of the markup, so having an `alt` should be the job of a package like `jest-axe` for example. Here, if the question is "Is this image labelled?". For me the question is yes. It is not optimal, it is not covering all usecases covered by `alt` *but* it is labelled.

**Comment by alexkrolick** [comment]
Hmm. Just throwing it out there - instead of having partially-correct rules with exceptions ("it is labelled but needs a src"), could we audit the whole DOM using the same rules as the `jsx-a11y` eslint plugin?

**Comment by hiwelo** [comment]
In this case, I would push people to use `jest-axe` or `react-axe`. React itself is [pushing people toward this solution](https://reactjs.org/docs/accessibility.html#testing-accessibility-in-the-browser).
Testing the accessibility of the whole DOM is something incredibly difficult especially because of the lack of alignment on a lot of question between the HTML5.1 guidelines, the WHATWG DOM implementation, the WAI-ARIA guidelines and the WCAG 2.1 rules.

In our case, could we have a different solution involving another kind of messages for `toBeLabelled`. This matcher could return for an image:

- no error if correctly labelled (image using `alt`)
- an error saying that "You should consider using `alt`" when `aria-label` is used on an image with or without `alt`.
- an error saying that nothing is labelled when everything is missing.


Could it be a good way to deal with this?

**Comment by Acventor** [comment]
In my case I change `aria-label` of button on click on it. It would be great to have matcher like `toBeLabelled `.

**Comment by gnapse** [comment]
Thanks for the reminder. I'll be working on this soon.
